indra/llcommon/lua_function.cpp

```diff
@@ -0,0 +1,894 @@
+LuaState::LuaState(script_finished_fn cb):
+    mCallback(cb),
+    mState(nullptr)
+{
+    initLuaState();
+}
+
+void LuaState::initLuaState() 
+{
+    if (mState)
+    {
+        lua_close(mState);
+    }
+    mState = luaL_newstate();
+    luaL_openlibs(mState);
+    LuaFunction::init(mState);
+    // Try to make print() write to our log.
+    lua_register(mState, "print", LuaFunction::get("print_info"));
+    // We don't want to have to prefix require().
+    lua_register(mState, "require", LuaFunction::get("require"));
+}
+
+LuaState::~LuaState()
+{
+    // Did somebody call obtainListener() on this LuaState?
+    // That is, is there a LuaListener key in its registry?
+    auto listener{ getListener() };
+    if (listener)
+    {
+        // if we got a LuaListener instance, destroy it
+        auto lptr{ listener.get() };
+        listener.reset();
+        delete lptr;
+    }
+
+    lua_close(mState);
+
+    if (mCallback)
+    {
+        // mError potentially set by previous checkLua() call(s)
+        mCallback(mError);
+    }    
+}
+
+bool LuaState::checkLua(const std::string& desc, int r)
+{
+    if (r != LUA_OK)
+    {
+        mError = lua_tostring(mState, -1);
+        lua_pop(mState, 1);
+
+        LL_WARNS() << desc << ": " << mError << LL_ENDL;
+        return false;
+    }
+    return true;
+}
+
+std::pair<int, LLSD> LuaState::expr(const std::string& desc, const std::string& text)
+{
+    if (! checkLua(desc, lluau::dostring(mState, desc, text)))
+        return { -1, mError };
+
+    // here we believe there was no error -- did the Lua fragment leave
+    // anything on the stack?
+    std::pair<int, LLSD> result{ lua_gettop(mState), {} };
+    if (result.first)
+    {
+        // aha, at least one entry on the stack!
+        if (result.first == 1)
+        {
+            // Don't forget that lua_tollsd() can throw Lua errors.
+            try
+            {
+                result.second = lua_tollsd(mState, 1);
+            }
+            catch (const std::exception& error)
+            {
+                // lua_tollsd() is designed to be called from a lua_function(),
+                // that is, from a C++ function called by Lua. In case of error,
+                // it throws a Lua error to be caught by the Lua runtime. expr()
+                // is a peculiar use case in which our C++ code is calling
+                // lua_tollsd() after return from the Lua runtime. We must catch
+                // the exception thrown for a Lua error, else it will propagate
+                // out to the main coroutine and terminate the viewer -- but since
+                // we instead of the Lua runtime catch it, our lua_State retains
+                // its internal error status. Any subsequent lua_pcall() calls
+                // with this lua_State will report error regardless of whether the
+                // chunk runs successfully. Get a new lua_State().
+                initLuaState();
```

Comment:
I'd argue for some LL_WARNS here and down there (533)

Reply:
Great minds! I added one [here](https://github.com/secondlife/viewer/blob/release/luau-scripting/indra/llcommon/lua_function.cpp#L517) and [down there](https://github.com/secondlife/viewer/blob/release/luau-scripting/indra/llcommon/lua_function.cpp#L546-L547) with my most recent PR.